File: Core/src/main/java/com/darwinreforged/server/core/types/tuple/Tuple.java
package com.darwinreforged.server.core.types.tuple;

/**
 * A tuple of objects.
 *
 * @param <K> The key
 * @param <V> The value
 */
public class Tuple<K, V> {

    /**
     * Creates a new {@link Tuple} with the desired {@code first} and
     * {@code second} objects.
     *
     * @param first The first object
     * @param second The second object
     * @param <K> The type of first object
     * @param <V> The type of second object
     * @return The new Tuple
     */
    public static <K, V> Tuple<K, V> of(K first, V second) {
        return new Tuple<>(first, second);
    }

    protected final K first;
    protected final V second;

    /**
     * Creates a new {@link Tuple}.
     *
     * @param first The first object
     * @param second The second object
     */
    public Tuple(K first, V second) {
        this.first = first;
        this.second = second;
    }

    /**
     * Gets the first object, otherwise known as "key".
     *
     * @return The first object
     */
    public K getFirst() {
        return this.first;
    }

    /**
     * Gets the second object, otherwise known as "value".
     *
     * @return The value
     */
    public V getSecond() {
        return this.second;
    }

    @SuppressWarnings("rawtypes")
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        final Tuple other = (Tuple) obj;
        return this.first.equals(other.first)
                && this.second.equals(other.second);
    }
}

File: core/src/main/kotlin/org/dockbox/darwin/core/objects/tuple/Tuple.kt
package org.dockbox.darwin.core.objects.tuple

class Tuple<K, V>(var first: K, var second: V) {

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || javaClass != other.javaClass) {
            return false
        }
        val other = other as Tuple<*, *>
        return first == other.first && second == other.second
    }

    override fun hashCode(): Int {
        var result = first?.hashCode() ?: 0
        result = 31 * result + (second?.hashCode() ?: 0)
        return result
    }

    companion object {
        fun <K, V> of(first: K, second: V): Tuple<K, V> {
            return Tuple(first, second)
        }
    }

}
